### PR TITLE
Add avatar

### PR DESCRIPTION
**Pull request by revonateB0T**

### CODE REVIEW

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+import moe.tarsin.coroutines.runSuspendCatching
+
+private val limitScope = CoroutineScope(Dispatchers.IO)
+private val refreshEvent = MutableSharedFlow<Unit>()
+
+private val limitFlow = refreshEvent.map { EhEngine.getImageLimits().right() }
+    .retryWhen<Either<String, HomeParser.Result>> { cause, _ ->
+        emit(cause.displayString().left())
+        delay(15.seconds)
+        true
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:65]: Retry does not work here.

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+@Composable
+fun AvatarIcon() {
+    val hasSignedIn by Settings.hasSignedIn.collectAsState()
+    val avatar by Settings.avatar.collectAsState()
+    if (hasSignedIn) {
+        val placeholder = stringResource(id = R.string.please_wait)
+        val resetImageLimitSucceed = stringResource(id = R.string.reset_limits_succeed)
+        val result by limitFlow.collectAsState(none())
+        LaunchedEffect(Unit) {
+            refreshEvent.emit(Unit)
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:78]: This should be in `onClick`.

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+            refreshAccountInfo()
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:79]: Changing avatar is expected to be rare, so we should only refresh it if the user initiated it.

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        emit(cause.displayString().left())
+        delay(15.seconds)
+        true
+    }.map { it.some() }.shareIn(limitScope, SharingStarted.Eagerly, replay = 1)
+
+context(CoroutineScope, DialogState, SnackbarHostState)
+@Composable
+fun AvatarIcon() {
+    val hasSignedIn by Settings.hasSignedIn.collectAsState()
+    val avatar by Settings.avatar.collectAsState()
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:72]: This can be moved inside `if`.

Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt]: Done

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt (hunk 3 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt]: Done

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt (hunk 2 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt]: Done

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt (hunk 1 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:63]: What do you mean?

Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt]: Done

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+import kotlinx.coroutines.launch
+import moe.tarsin.coroutines.runSuspendCatching
+
+private val limitScope = CoroutineScope(Dispatchers.IO)
+private val refreshEvent = MutableSharedFlow<Unit>()
+private suspend fun refresh() = refreshEvent.emit(Unit)
+private fun <T> Flow<T>.keepIf(predicate: (old: T, new: T) -> Boolean) = distinctUntilChanged(predicate)
+
+private val limitFlow = refreshEvent.conflate().map { EhEngine.getImageLimits().right() }
+    .retryWhen<Either<String, HomeParser.Result>> { e, _ -> emit(e.displayString().left()).let { true } }
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:67]: Why not use `Either.catch().mapLeft()`?

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    .map(::Some).let { merge(it, refreshEvent.map { none() }) }
+    .keepIf { o, n -> o.isSome { it.isRight() } && n.isNone() }
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:69]: We want to show progress indicator when the user resets the limit, that will be filtered by this.

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+        val resetImageLimitSucceed = stringResource(id = R.string.reset_limits_succeed)
+        val result by limitFlow.collectAsState()
+        IconButton(
+            onClick = {
+                launch {
+                    refresh()
+                    awaitConfirmationOrCancel(
+                        confirmText = R.string.reset,
+                        title = R.string.image_limits,
+                        confirmButtonEnabled = result.isSome { it.isRight { it.limits.resetCost != 0 } },
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:87]: If result changes while the dialog is shown, this won't change.
We should move the reset button into the dialog.

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+                                            type = "C",
+                                            amount = funds.credit,
+                                        )
+                                    }
+                                }
+                            }
+                        }
+                    }
+                    runSuspendCatching {
+                        EhEngine.resetImageLimits()
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:140]: We should show the error if it fails.

File: app/src/main/kotlin/com/hippo/ehviewer/ui/settings/EhScreen.kt [hunk 9]
```diff
@@ -100,29 +79,28 @@ fun EhScreen(navigator: DestinationsNavigator) {
...
-                        showCancelButton = cookies.last().second != null,
-                        onCancelButtonClick = { EhCookieStore.clearIgneous() },
-                        secure = hasSignedIn,
-                    ) {
-                        if (hasSignedIn) {
+            if (hasSignedIn) {
+                val displayName by Settings.displayName.collectAsState()
+                Preference(
+                    title = stringResource(id = R.string.account_name),
+                    summary = displayName ?: "???",
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/settings/EhScreen.kt:93]: "???" is actually a valid user name.

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt (hunk 5 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt]: Done

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt (hunk 6 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt]: Done

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt (hunk 7 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt]: Done

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt (hunk 8 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:167]: Done

File: app/src/main/kotlin/com/hippo/ehviewer/ui/settings/EhScreen.kt (hunk 9 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/settings/EhScreen.kt:93]: Suggestion?

Comment (FooIbar) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/settings/EhScreen.kt:93]: Just leave it to null?

Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/settings/EhScreen.kt]: Done

File: app/src/main/kotlin/com/hippo/ehviewer/ui/tools/Dialog.kt [hunk 10]
```diff
@@ -281,14 +281,32 @@ class DialogState {
         onCancelButtonClick: () -> Unit = {},
         secure: Boolean = false,
         text: @Composable (() -> Unit)? = null,
+    ) = awaitConfirmationOrCancel(
+        confirmButton = {
+            TextButton(onClick = it, enabled = confirmButtonEnabled) {
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/tools/Dialog.kt:286]: The `confirmButtonEnabled` parameter can be removed as the new overload is added.
Though if we want to move clear igneous to the dialog in the future, it's better to use a no button dialog and place the reset button inside.

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+fun AvatarIcon() {
+    val hasSignedIn by Settings.hasSignedIn.collectAsState()
+    if (hasSignedIn) {
+        val placeholder = stringResource(id = R.string.please_wait)
+        val resetImageLimitSucceed = stringResource(id = R.string.reset_limits_succeed)
+        val result by limitFlow.collectAsState()
+        IconButton(
+            onClick = {
+                launch {
+                    refreshEvent.emit(Unit)
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:101]: This is unnecessary now.

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+    }
+    .map { catch { EhEngine.getImageLimits() }.mapLeft { e -> e.displayString() } }
+    .let { src -> merge(src.map { v -> v.some() }, invalidateEvent.map { none() }) }
+    .stateIn(limitScope, SharingStarted.Eagerly, none())
+    .apply {
+        limitScope.launch {
+            collectLatest {
+                // Consider cached value needs invalidate for 2 minutes
+                delay(120.seconds)
+                invalidateEvent.emit(Unit)
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:85]: It works but feels clunky.

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt (hunk 11 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:101]: This is necessary as user may want see updated value, but cache is not invalidated

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+
+private val limitScope = CoroutineScope(Dispatchers.IO)
+private val refreshEvent = MutableSharedFlow<Unit>()
+private val invalidateEvent = MutableSharedFlow<Unit>()
+
+typealias Result = Option<Either<String, HomeParser.Result>>
+
+private val limitFlow: StateFlow<Result> = refreshEvent.conflate()
+    .map { catch { EhEngine.getImageLimits() }.mapLeft { e -> e.displayString() } }
+    .let { src -> merge(src.map { v -> v.some() }, invalidateEvent.map { none() }) }
```
[the remark this answers is not in the document]
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt]: Done

File: app/src/main/kotlin/com/hippo/ehviewer/ui/tools/Dialog.kt (hunk 10 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/tools/Dialog.kt]: Done

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt (hunk 12 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:85]: Maybe we just not invalidate cache, and bonus animated quota

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt (hunk 11 above)
Comment (FooIbar) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:89]: Then what's the cache for? It doesn't actually reduce network requests.

Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:89]: For further pretty animation :)

File: app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+        val result by limitFlow.collectAsState()
+        IconButton(
+            onClick = {
+                launch {
+                    refreshEvent.emit(Unit)
+                    awaitConfirmationOrCancel(
+                        confirmButton = {
+                            TextButton(onClick = it, enabled = result.isSome { it.isRight { it.limits.resetCost != 0 } }) {
+                                Text(text = stringResource(id = R.string.reset))
+                            }
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:94]: Resetting image limit should not close the dialog as the dialog is the only place to show image limit now.

Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:96]: Then should we show snackbar?

Comment (FooIbar) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:109]: Only on failure to show the reason? The user can tell from the new image limit if it succeeded.

Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/main/AvatarIcon.kt:109]: Done

**Review by FooIbar — APPROVED** [review]

**Pull request merged**